tests/Console/ModelsCommand/CustomDate/__snapshots__/Test__test__1.php

```diff
@@ -14,8 +14,8 @@
  * @method static \Illuminate\Database\Eloquent\Builder|CustomDate newModelQuery()
  * @method static \Illuminate\Database\Eloquent\Builder|CustomDate newQuery()
  * @method static \Illuminate\Database\Eloquent\Builder|CustomDate query()
- * @method static \Illuminate\Database\Eloquent\Builder|CustomDate whereCreatedAt($value)
- * @method static \Illuminate\Database\Eloquent\Builder|CustomDate whereUpdatedAt($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|CustomDate whereCreatedAt(\Carbon\CarbonImmutable|null $value)
```

Comment:
I mention this in my previous feedback but it might just have gotten lost: any method accepting Carbon (or a variation thereof) also accept strings, it doesn't  just accept an instance (or null).

Reply:
Understood, but if you look at the created_at property, it's also `\Carbon\CarbonImmutable|null`, so this is not an issue with this new feature.

Reply:
The `whereProperty` method parameter type is now exactly the same as the corresponding `@property`.

Reply:
> Understood, but if you look at the created_at property, it's also `\Carbon\CarbonImmutable|null`, so this is not an issue with this new feature.
…
> The `whereProperty` method parameter type is now exactly the same as the corresponding `@property`.

I pointed this out in https://github.com/barryvdh/laravel-ide-helper/pull/989#pullrequestreview-447800637
> Basically it assumes that the "cast output" is the sole allowed input for those magic methods.

However this assumption isn't correct.

Let's look at this tinker session:
```
$ ./artisan tinker
Psy Shell v0.10.4 (PHP 7.4.10 — cli) by Justin Hileman
>>> $user = App\Models\User;
=> App\Modes\User {#4302
   }
>>> $user->deleted_at;
=> null
>>> $user->deleted_at = '2020-10-10 12:12:00';
=> "2020-10-10 12:12:00"
>>> $user->deleted_at
=> Carbon\CarbonImmutable @1602331920 {#4301
     date: 2020-10-10 12:12:00.0 UTC (+00:00),
   }
>>> $user->deleted_at = \Carbon\CarbonImmutable::now();
=> Carbon\CarbonImmutable @1599856846 {#4319
     date: 2020-09-11 20:40:46.689963 UTC (+00:00),
   }
>>> $user->deleted_at
=> Carbon\CarbonImmutable @1599856846 {#4309
     date: 2020-09-11 20:40:46.0 UTC (+00:00),
   }
```

Setting either string **or** Carbon is fine.

However, in the IDE:
![image](https://user-images.githubusercontent.com/87493/92971070-13004f00-f480-11ea-92f4-5a5541550f81.png)

----
I noticed something else in a private project I tested this:

Migration:
```php
            $table->jsonb('keywords')->nullable();
```
Model:
```php
 * @property array|null $keywords
…
    protected $casts = [
        'keywords' => 'array',
    ];
```
Before:
```php
 * @method static Builder|Automation whereKeywords($value)
```
After:
```php
 * @method static Builder|Automation whereKeywords(string|null $value)
```

I think `string` is practical in this context, there's not much point about this anyway as it's a JSON column.

----
But then I've another model with a similar column which however generated this:

Migration:
```php
            $table->json('some_other_column')->nullable();
```
Model:
```php
 * @property array $some_other_column
…
 * @method static Builder|Attachment whereSomeOtherColumn(mixed|null $value)
…
    protected $casts = [
        'some_other_column' => 'array',
    ];
…
    public function setSomeOtherColumnAttribute(?array $value): self
    {
…    }

    public function getSomeOtherColumnAttribute(): array
    {
…    }
```
The `mixed` is in fact also not that bad here, actually might make even more sense.

Though I could not figure out why one is `string` here and one is `mixed`. I tried removing the set/get but it didn't change 🤷‍♀️

----
Whilst I'm not yet in the clear regarding the latter new examples, I think that date casts not accepting string is bad. The `laravel-ide-helper` is supposed to help but would get in the way here.

These were only easy examples with `$date` and `array` casts, but maybe properties in `$casts` have to be handled differently?

Reply:
I'm not disputing the string issue with dates. However, that's an existing issue in laravel-ide-helper. You'd have the same issue with setting the magic properties. I'm not going to fix this issue in this PR. That's outside the scope of the PR. If you want that fixed, then create an issue and maybe someone will create a PR to rectify the issue.

Reply:
>  I'm not going to fix this issue in this PR. That's outside the scope of the PR. If you want that fixed, then create an issue and maybe someone will create a PR to rectify the issue.

That's a pretty strong feedback here, not sure where this is coming from.

----
It made me think though whether this whole feature should be optional  (opt-in/out => debatable)

Reply:
Using the example of dates as `\Carbon\CarbonImmutable` and not allowing `string`s is an issue that already exists in the code base. I did not introduce this issue. If you look at the diffs for the test snapshots, you'll see that the `@property` tags did not change.

My objection to your feedback is that you're asking me to fix an issue, an issue that already existed, when I'm trying to introduce a new feature. If this existing issue stood in the way of introducing this new feature, then I would fix it. However, the issue does not hinder this feature's implementation. Clearly, what this feature does do is highlight the existing issue. I agree the issue should be fixed, but I believe that it should be fixed in a new PR, one that targets the issue specifically.

I don't have any objection to adding a flag for this feature.

Reply:
I created a new PR (#1056) to fix the date type.

Reply:
One more thing: looking at the JSON example you gave, I will try to fix that one. That has to do with the command not checking casts until after getting properties from the table:

```php
// ModelsCommand::generateDocs
                    if ($hasDoctrine) {
                        $this->getPropertiesFromTable($model);
                    }

                    if (method_exists($model, 'getCasts')) {
                        $this->castPropertiesType($model);
                    }
```

`castPropertiesType` should probably be called from `getPropertiesFromTable`.